components/concurrency_manager/src/lib.rs

```diff
@@ -52,9 +75,61 @@ impl ConcurrencyManager {
 
     /// Updates max_ts with the given new_ts. It has no effect if
     /// max_ts >= new_ts or new_ts is TimeStamp::max().
-    pub fn update_max_ts(&self, new_ts: TimeStamp) {
-        if new_ts != TimeStamp::max() {
-            self.max_ts.fetch_max(new_ts.into_inner(), Ordering::SeqCst);
+    ///
+    /// To avoid invalid ts breaking the invariants, the new_ts should be
+    /// less than or equal to the max_ts_limit.
+    ///
+    /// # Returns
+    /// - Ok(()): If the update is successful or has no effect
+    /// - Err(limit): If new_ts is greater than the max_ts_limit, returns the
+    ///   current limit value
+    pub fn update_max_ts(
+        &self,
+        new_ts: TimeStamp,
+        source: Option<String>,
+    ) -> Result<(), InvalidMaxTsUpdate> {
+        if new_ts == TimeStamp::max() {
+            return Ok(());
+        }
+        let new_ts = new_ts.into_inner();
+        let limit = self.max_ts_limit.load(Ordering::SeqCst);
+        if limit > 0 && new_ts > limit {
```

Comment:
Should we consider the case that PD and TiKV is network-isolated and the limit is not updated in time? How about adding an additional gap according to the duration that has been elapsed since the last successful `set_max_ts_limit` call? We can first check `new_ts > limit` as the fast path, and check `new_ts > limit + gap` then if the former failed, so that the overhead of getting the system time can be avoided in most cases.

Reply:
Yeah already added just as your suggestion 😁. PTAL